pkg/network/common/egress_dns.go

```diff
@@ -116,39 +143,41 @@ func (e *EgressDNS) Sync() {
 	}
 }
 
-func (e *EgressDNS) GetMinQueryTime() (time.Time, ktypes.UID, string, bool) {
+func (e *EgressDNS) GetMinQueryTime() (time.Time, string, []PolicyUpdate, bool) {
 	e.lock.Lock()
 	defer e.lock.Unlock()
+	policyUpdates := make([]PolicyUpdate, 0)
+	tm, dnsName, timeSet := e.dns.GetMinQueryTime()
+	if !timeSet {
+		klog.Warningf("Time was not set for dns name: %s", dnsName)
+		return tm, dnsName, nil, timeSet
+	}
 
-	timeSet := false
-	var minTime time.Time
-	var uid ktypes.UID
-
-	for policyUID, dnsInfo := range e.pdMap {
-		tm, ok := dnsInfo.GetMinQueryTime()
-		if !ok {
-			continue
-		}
-
-		if (timeSet == false) || tm.Before(minTime) {
-			timeSet = true
-			minTime = tm
-			uid = policyUID
+	if uids, ok := e.dnsNamesToUIDsMap[dnsName]; ok {
+		for _, uid := range uids {
+			policyUpdates = append(policyUpdates, PolicyUpdate{uid, e.namespaces[uid]})
 		}
+	} else {
+		klog.V(5).Infof("Didn't find any entry for dns name: %s in the dns map.", dnsName)
 	}
-
-	return minTime, uid, e.namespaces[uid], timeSet
+	return tm, dnsName, policyUpdates, timeSet
 }
 
 func (e *EgressDNS) GetIPs(policy networkv1.EgressNetworkPolicy, dnsName string) []net.IP {
 	e.lock.Lock()
 	defer e.lock.Unlock()
 
-	dnsInfo, ok := e.pdMap[policy.UID]
-	if !ok {
+	if uids, ok := e.dnsNamesToUIDsMap[dnsName]; ok {
+		for _, uid := range uids {
+			if uid == policy.UID {
+				return e.dns.Get(dnsName).ips
```

Comment:
This is weird. Just change `GetIPs` to no longer take a `policy`.

Reply:
I thought about it. The only reason I am still passing in the `policy` is to make sure we don't return valid IPs for a policy that doesn't have the corresponding `dnsName` in its rule set. So in some ways it is checking that the GetIPs only returns valid ips for policies that indeed have the dns name in the egress rules. I am OK with removing it as well if that additional check is not needed semantically.

Reply:
The only reason it was taking a policy before though was because it *needed* to in order to look up the information. The thing that the caller wants is the name-to-IP mapping. Being forced to pass a policy was just an implementation detail.